File: src/main/java/me/kansio/client/modules/impl/movement/Flight.java
package me.kansio.client.modules.impl.movement;

import dorkbox.messageBus.annotations.Subscribe;
import lombok.Getter;
import me.kansio.client.event.impl.BlockCollisionEvent;
import me.kansio.client.event.impl.MoveEvent;
import me.kansio.client.event.impl.PacketEvent;
import me.kansio.client.event.impl.UpdateEvent;
import me.kansio.client.modules.api.ModuleCategory;
import me.kansio.client.modules.api.ModuleData;
import me.kansio.client.modules.impl.Module;
import me.kansio.client.modules.impl.movement.flight.FlightMode;
import me.kansio.client.property.value.BooleanValue;
import me.kansio.client.property.value.ModeValue;
import me.kansio.client.property.value.NumberValue;
import me.kansio.client.utils.java.ReflectUtils;
import me.kansio.client.utils.math.Stopwatch;
import net.minecraft.potion.Potion;

import java.util.Comparator;
import java.util.List;
import java.util.stream.Collectors;

@ModuleData(
        name = "Flight",
        category = ModuleCategory.MOVEMENT,
        description = "Allows you to fly"
)
public class Flight extends Module {

    private final List<? extends FlightMode> modes = ReflectUtils.getReflects(this.getClass().getPackage().getName() + ".flight", FlightMode.class).stream()
            .map(aClass -> {
                try {
                    return aClass.getDeclaredConstructor().newInstance();
                } catch (Exception e) {
                    e.printStackTrace();
                }
                return null;
            })
            .sorted(Comparator.comparing(speedMode -> speedMode != null ? speedMode.getName() : null))
            .collect(Collectors.toList());

    private final ModeValue modeValue = new ModeValue("Mode", this, modes.stream().map(FlightMode::getName).collect(Collectors.toList()).toArray(new String[]{}));
    private FlightMode currentMode = modes.stream().anyMatch(speedMode -> speedMode.getName().equalsIgnoreCase(modeValue.getValue())) ? modes.stream().filter(speedMode -> speedMode.getName().equalsIgnoreCase(modeValue.getValue())).findAny().get() : null;
    @Getter private NumberValue<Double> speed = new NumberValue<>("Speed", this, 1d, 0d, 10d, 0.1);
    @Getter private BooleanValue viewbob = new BooleanValue("View Bobbing", this, true);
    @Getter private BooleanValue boost = new BooleanValue("Boost", this, true, modeValue, "Funcraft");
    @Getter private BooleanValue extraBoost = new BooleanValue("Extra Boost", this, true, modeValue, "Funcraft");
    @Getter private BooleanValue glide = new BooleanValue("Glide", this, true, modeValue, "Funcraft");
    @Getter private ModeValue boostMode = new ModeValue("Boost Mode", this, modeValue, new String[]{"Funcraft"}, "Normal", "Damage", "WOWOMG");
    @Getter private NumberValue<Double> timer = new NumberValue<>("Timer", this, 1d, 1d, 5d, 0.1, modeValue, "Mush");

    private Stopwatch stopwatch = new Stopwatch();

    public float ticks = 0;
    public float prevFOV = mc.gameSettings.fovSetting;
    private int level;
    private double moveSpeed, lastDist;

    public void onEnable() {
        this.currentMode = modes.stream().anyMatch(speedMode -> speedMode.getName().equalsIgnoreCase(modeValue.getValue())) ? modes.stream().filter(speedMode -> speedMode.getName().equalsIgnoreCase(modeValue.getValue())).findAny().get() : null;
        currentMode.onEnable();
    }

    public void onDisable() {
        mc.thePlayer.motionX = 0;
        mc.thePlayer.motionY = 0;
        mc.thePlayer.motionZ = 0;
        mc.timer.timerSpeed = 1f;
        currentMode.onDisable();
    }

    @Subscribe
    public void onUpdate(UpdateEvent event) {
        if (viewbob.getValue() && mc.thePlayer.isMoving()) {
            mc.thePlayer.cameraYaw = 1;
            //mc.thePlayer.cameraYaw = 0.1f;
            //mc.thePlayer.cameraYaw = 1000f;
        } else {
            mc.thePlayer.cameraYaw = 0;
        }
        currentMode.onUpdate(event);
    }

    @Subscribe
    public void onMove(MoveEvent event) {
        currentMode.onMove(event);
    }

    @Subscribe
    public void onPacket(PacketEvent event) {
        currentMode.onPacket(event);
    }

    @Subscribe
    public void onCollide(BlockCollisionEvent event) {
        currentMode.onCollide(event);
    }

    private double getBaseMoveSpeed() {
        double n = 0.2873;
        if (mc.thePlayer.isPotionActive(Potion.moveSpeed)) {
            n *= 1.0 + 0.2 * (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).getAmplifier() + 1);
        }
        return n;
    }

    @Override
    public String getSuffix() {
        return " " + modeValue.getValueAsString();
    }
}

File: src/main/java/me/kansio/client/modules/impl/movement/Flight.kt
package me.kansio.client.modules.impl.movement

import me.kansio.client.modules.api.ModuleData
import me.kansio.client.modules.api.ModuleCategory
import me.kansio.client.modules.impl.movement.flight.FlightMode
import me.kansio.client.utils.java.ReflectUtils
import java.util.stream.Collectors
import me.kansio.client.value.value.ModeValue
import me.kansio.client.value.value.NumberValue
import com.google.common.eventbus.Subscribe
import lombok.Getter
import me.kansio.client.event.impl.UpdateEvent
import me.kansio.client.gui.notification.NotificationManager
import me.kansio.client.event.impl.MoveEvent
import me.kansio.client.event.impl.PacketEvent
import me.kansio.client.event.impl.BlockCollisionEvent
import me.kansio.client.gui.notification.Notification
import me.kansio.client.modules.impl.Module
import me.kansio.client.utils.math.Stopwatch
import me.kansio.client.utils.player.TimerUtil
import me.kansio.client.value.value.BooleanValue
import net.minecraft.potion.Potion
import java.lang.Exception
import java.util.Comparator

@Getter
@ModuleData(name = "Flight", category = ModuleCategory.MOVEMENT, description = "Allows you to fly")
class Flight : Module() {
    private val modes = ReflectUtils.getReflects("${this.javaClass.`package`.name}.flight", FlightMode::class.java).map { it.newInstance() as FlightMode }.sortedBy { it.name }
    val currentMode: FlightMode get() = modes.find { mode.equals(it.name) } ?: throw NullPointerException() // this should not happen
    private val mode = ModeValue(
        "Mode",
        this,
        *modes.map { it.name }.toTypedArray()
    )

    val speed = NumberValue("Speed", this, 1.0, 0.0, 10.0, 0.1)
    val antikick = BooleanValue("AntiKick", this, true, mode, "BridgerLand (TP)")
    val timer = NumberValue("Timer", this, 1.0, 1.0, 5.0, 0.1, mode, "Mush")
    val blink = BooleanValue("Blink", this, true, mode, "Mush")
    val viewbob = BooleanValue("View Bobbing", this, true)
    private val stopwatch = Stopwatch()
    var ticks = 0f
    var prevFOV = 0f
    override fun onEnable() {
        currentMode.onEnable()
    }

    override fun onDisable() {
        mc.thePlayer.motionX = 0.0
        mc.thePlayer.motionY = 0.0
        mc.thePlayer.motionZ = 0.0
        TimerUtil.Reset()
        currentMode.onDisable()
    }

    @Subscribe
    fun onUpdate(event: UpdateEvent?) {
        if (viewbob.value && mc.thePlayer.isMoving) {
            mc.thePlayer.cameraYaw = 0.05f
        } else {
            mc.thePlayer.cameraYaw = 0f
        }
        currentMode.onUpdate(event)
        if (mc.thePlayer.ticksExisted < 5) {
            if (isToggled) {
                toggle()
            }
        }
    }

    @Subscribe
    fun onMove(event: MoveEvent?) {
        currentMode.onMove(event)
    }

    @Subscribe
    fun onPacket(event: PacketEvent?) {
        currentMode.onPacket(event)
    }

    @Subscribe
    fun onCollide(event: BlockCollisionEvent?) {
        currentMode.onCollide(event)
    }

    private val baseMoveSpeed: Double
        private get() {
            var n = 0.2873
            if (mc.thePlayer.isPotionActive(Potion.moveSpeed)) {
                n *= 1.0 + 0.2 * (mc.thePlayer.getActivePotionEffect(Potion.moveSpeed).amplifier + 1)
            }
            return n
        }

    override fun getSuffix(): String {
        return " " + mode.valueAsString
    }
}